### PR TITLE
Add simple Future library

### PR DESCRIPTION
**Pull request by graetzer**
Pleae read [the guide](https://github.com/arangodb/arangodb/blob/feature/async-futures/arangod/Futures/README.md) for more information

### CODE REVIEW

**Comment by graetzer** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/1061/

File: arangod/Scheduler/Scheduler.cpp [hunk 1]
```diff
@@ -266,31 +266,30 @@ void Scheduler::post(std::function<void()> const callback, bool isV8,
       std::shared_ptr<asio_ns::deadline_timer> timer(
           newDeadlineTimer(boost::posix_time::millisec(timeout)));
       timer->async_wait(
-          [this, callback, isV8, t, timer](const asio::error_code& error) {
+        [this, cb = std::move(cb), isV8, t, timer](const asio::error_code& error) {
```
Comment (mpoeter) [inline arangod/Scheduler/Scheduler.cpp]: Yay! Finally we can use generalized lambda captures! ❤️

**Comment by mpoeter** [comment]
Nice! But I have been wondering for some time now which programming model would be better suited for ArangoDB - futures or async/finish (see [Variations on task scheduling for shared memory system](http://repositum.tuwien.ac.at/obvutwhs/download/pdf/1633888?originalFilename=true) Section 2.2 for various programming models based on task parallelism).

**Comment by graetzer** [comment]
we need something that is easy to integrate into our existing codebase. I think Futures are suitable for this, they also don't require language extensions

File: arangod/Futures/SharedState.h [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+      switch (state) {
+        case State::Start:
+          if (_state.compare_exchange_strong(state, State::OnlyResult, std::memory_order_release)) {
+            return;
+          }
+          TRI_ASSERT(state == State::OnlyCallback); // race with setCallback
+          [[gnu::fallthrough]];
+          
+        case State::OnlyCallback:
+          if (_state.compare_exchange_strong(state, State::Done, std::memory_order_release)) {
```
Comment (mpoeter) [inline arangod/Futures/SharedState.h]: This CAS has to use memory_order_acquire to synchronize-with the release-CAS in line 166.

File: arangod/Futures/SharedState.h [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+      switch (state) {
+        case State::Start:
+          if (_state.compare_exchange_strong(state, State::OnlyCallback, std::memory_order_release)) {
+            return;
+          }
+          TRI_ASSERT(state == State::OnlyResult); // race with setResult
+          [[gnu::fallthrough]];
+          
+        case State::OnlyResult:
+          if (_state.compare_exchange_strong(state, State::Done, std::memory_order_release)) {
```
Comment (mpoeter) [inline arangod/Futures/SharedState.h]: This CAS has to use memory_order_acquire to synchronize-with the release-CAS in line 202.

**Comment by mpoeter** [comment]
Cilk may come with some language extensions, but async/finish can easily be implemented in a library as well. I have been planning to introduce task based parallelism into ArangoDB for some time now, but I am still not sure which programming model would be best suited - mostly because I only have a very limited knowledge of the code and the overall architecture, so you are certainly in a better position to judge that.

But this is not a strict either-or decision on futures vs. async/finish - both models can happily coexist, or even be combined.

**Comment by mpoeter** [comment]
More on the actual topic of this PR: do you plan to add a function like `std::async` that returns a future as well?

**Comment by graetzer** [comment]
not sure maybe as Syntax sugar instead of using the scheduler code.

**Comment by mpoeter** [comment]
Well, IMHO there is conceptual difference. The scheduler supports a simple fire-and-forget model. But when I create a future I want to be able to access the result at a later time. And in case the future was not yet picked up by some other thread, the thread that is requesting the result should itself claim ownership and perform the necessary action to produce that result. So this functionality may build upon the scheduler, but it requires a little more than just posting a job.

**Comment by jsteemann** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/1119/

File: arangod/Futures/Future.h [hunk 4]
```diff
@@ -0,0 +1,557 @@
...
+#ifndef ARANGOD_FUTURES_FUTURE_H
+#define ARANGOD_FUTURES_FUTURE_H 1
+
+#include <chrono>
+#include <future>
+#include <thread>
+
+#include "Futures/Exceptions.h"
+#include "Futures/Unit.h"
+#include "Futures/Promise.h"
```
Comment (ObiWahn) [inline arangod/Futures/Future.h]: `Future.h` includes `Promise.h`

File: arangod/Futures/Promise.h [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+  
+private:
+  detail::SharedState<T>* _state;
+  /// Whether the Future has been retrieved (a one-time operation)
+  bool _retrieved;
+};
+}}
+#endif // ARANGOD_FUTURES_PROMISE_H
+
+#include "Future.h"
```
Comment (ObiWahn) [inline arangod/Futures/Promise.h]: `Promise.h` includes `Future.h` -> cyclic dependency

The future`s include guard will never close.

File: arangod/Futures/Future.h (hunk 4 above)
Comment (graetzer) [reply arangod/Futures/Future.h]: But the second time the guard is defined, so it will stop.

File: arangod/Futures/Future.h [hunk 6]
```diff
@@ -0,0 +1,557 @@
...
+public:
+  
+  /// @brief Constructs a Future with no shared state.
+  static Future<T> makeEmpty() {
+    return Future<T>(detail::EmptyConstructor{});
+  }
+  
+  /// Construct a Future from a value (perfect forwarding)
+  template <
+  class T2 = T,
```
[the remark this answers is not in the document]
Comment (graetzer) [reply arangod/Futures/Future.h]: it should allow you to construct this with any value convertible to T

File: arangod/Scheduler/SchedulerFeature.cpp [hunk 7]
```diff
@@ -157,15 +157,16 @@ void SchedulerFeature::start() {
...
+  try {
+    auto* dealer = ApplicationServer::getFeature<V8DealerFeature>("V8Dealer");
+    if (dealer->isEnabled()) {
+      dealer->defineContextUpdate(
+          [](v8::Isolate* isolate, v8::Handle<v8::Context> context, size_t) {
+            TRI_InitV8Dispatcher(isolate, context);
+          },
+          nullptr);
+    }
+  } catch(...) {}
```
Comment (ObiWahn) [inline arangod/Scheduler/SchedulerFeature.cpp]: Is this change intentional? Is it part of the futures PR?

File: arangod/RocksDBEngine/RocksDBCommon.cpp [hunk 8]
```diff
@@ -133,6 +133,7 @@ std::size_t countKeyRange(rocksdb::DB* db,
 
   rocksdb::ReadOptions readOptions;
   readOptions.prefix_same_as_start = prefix_same_as_start;
+  readOptions.ignore_range_deletions = false;
```
Comment (ObiWahn) [inline arangod/RocksDBEngine/RocksDBCommon.cpp]: Is this part of the the futures lib?

File: arangod/Futures/backports.h [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+std::is_constructible<
+std::function<R(Args ...)>,
+std::reference_wrapper<typename std::remove_reference<F>::type>
+>
+{};
+  
+//  mimic: std::invoke, C++17
+template <typename F, typename... Args>
+constexpr auto invoke(F&& f, Args&&... args) noexcept(noexcept(static_cast<F&&>(f)(static_cast<Args&&>(args)...)))
+-> decltype(static_cast<F&&>(f)(static_cast<Args&&>(args)...)) {
```
Comment (ObiWahn) [inline arangod/Futures/backports.h]: Do you want to convert lvalue refs to rvalue refs?

Comment (graetzer) [reply arangod/Futures/backports.h]: yes its std::forward for variadic template args

File: arangod/Scheduler/SchedulerFeature.cpp (hunk 7 above)
Comment (graetzer) [reply arangod/Scheduler/SchedulerFeature.cpp]: yes, was needed to start the SchedulerFeature in catch tests

File: arangod/Futures/backports.h (hunk 9 above)
Comment (ObiWahn) [reply arangod/Futures/backports.h]: No it is move:)

**Comment by jsteemann** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/1208/

File: arangod/Aql/SharedQueryState.cpp [hunk 10]
```diff
@@ -29,71 +29,48 @@ using namespace arangodb;
...
+  std::lock_guard<std::mutex> guard(_mutex);
   _continueCallback = handler;
   _hasHandler = true;
 }
+
+/// execute the _continueCallback. must hold _mutex
+bool SharedQueryState::executeContinueCallback() const {
+  TRI_ASSERT(_hasHandler);
+  auto scheduler = SchedulerFeature::SCHEDULER;
+  TRI_ASSERT(scheduler != nullptr);
```
Comment (jsteemann) [inline arangod/Aql/SharedQueryState.cpp]: looks like this assert is wrong in case of a shutdown

File: arangod/Scheduler/Scheduler.cpp [hunk 11]
```diff
@@ -215,7 +215,7 @@ Scheduler::~Scheduler() {
 }
 
 // do not pass callback by reference, might get deleted before execution
```
Comment (jsteemann) [inline arangod/Scheduler/Scheduler.cpp]: if passing by reference is valid, the comment above should be adjusted.

File: lib/Futures/SharedState.h [hunk 12]
```diff
@@ -0,0 +1,301 @@
...
+////////////////////////////////////////////////////////////////////////////////
+
+#ifndef ARANGOD_FUTURES_SHARED_STATE_H
+#define ARANGOD_FUTURES_SHARED_STATE_H 1
+
+#include <atomic>
+
+#include "Futures/Try.h"
+#include "Futures/function2/function2.hpp"
+//#include "Scheduler/Scheduler.h"
```
Comment (jsteemann) [inline lib/Futures/SharedState.h]: can be removed
